presto-native-execution/presto_cpp/main/operators/LocalShuffle.cpp

```diff
@@ -319,22 +441,120 @@ LocalShuffleReader::LocalShuffleReader(
+          folly::join(", ", partitionIds_));
+      auto reader = std::make_unique<SortedShuffleFileStreamReader>(
+          filename, streamIndex, pool_);
+      if (reader->hasData()) {
+        streams.push_back(std::move(reader));
+        ++streamIndex;
+      }
+    }
+    if (!streams.empty()) {
+      merge_ =
+          std::make_unique<velox::TreeOfLosers<velox::MergeStream, uint16_t>>(
+              std::move(streams));
+    }
+
+    mergeInitialized_ = true;
+  }
+
+  auto batchBuffer = AlignedBuffer::allocate<char>(maxBytes, pool_, 0);
+  std::vector<std::string_view> rows;
+  size_t totalBytes = 0;
+  size_t bufferUsed = 0;
+
+  VELOX_CHECK(
+      merge_ || readPartitionFiles_.empty(),
+      "Failed to initialize merge: no valid shuffle files found");
+
+  if (merge_ == nullptr) {
+    return batches;
   }
 
+  while (auto* stream = merge_->next()) {
+    auto* reader = dynamic_cast<SortedShuffleFileStreamReader*>(stream);
+    const auto data = reader->currentData();
+    const size_t rowSize = kUint32Size + data.size();
+
+    // Ensure single row doesn't exceed maxBytes limit
+    VELOX_CHECK_LE(
+        rowSize,
+        maxBytes,
+        "Single row size {} bytes exceeds maxBytes limit {} bytes. "
+        "This indicates data corruption or misconfiguration.",
+        rowSize,
+        maxBytes);
+
+    // Check if adding this row would exceed maxBytes limit
+    if (totalBytes > 0 && totalBytes + rowSize > maxBytes) {
+      batches.push_back(
+          std::make_unique<ReadBatch>(std::move(rows), std::move(batchBuffer)));
+      return batches;
+    }
+
+    // Write row: [dataSize][data]
+    char* writePos = batchBuffer->asMutable<char>() + bufferUsed;
+    *reinterpret_cast<TRowSize*>(writePos) =
+        folly::Endian::big(static_cast<TRowSize>(data.size()));
+
+    if (!data.empty()) {
+      memcpy(writePos + sizeof(TRowSize), data.data(), data.size());
+    }
+
+    rows.emplace_back(batchBuffer->as<char>() + bufferUsed, rowSize);
+    bufferUsed += rowSize;
+    totalBytes += rowSize;
+
+    reader->next();
+  }
+
+  if (!rows.empty()) {
+    batches.push_back(
+        std::make_unique<ReadBatch>(std::move(rows), std::move(batchBuffer)));
+  }
+
+  return batches;
+}
+
+std::vector<std::unique_ptr<ReadBatch>> LocalShuffleReader::nextUnsorted(
+    uint64_t maxBytes) {
   std::vector<std::unique_ptr<ReadBatch>> batches;
   uint64_t totalBytes{0};
-  // Read files until we reach maxBytes limit or run out of files.
+
   while (readPartitionFileIndex_ < readPartitionFiles_.size()) {
     const auto filename = readPartitionFiles_[readPartitionFileIndex_];
     auto file = fileSystem_->openFileForRead(filename);
     const auto fileSize = file->size();
 
-    // Stop if adding this file would exceed maxBytes (unless we haven't read
-    // any files yet)
+    // Ensure single file doesn't exceed maxBytes limit
+    // This guarantees we can always make progress and prevents OOM
+    VELOX_CHECK_LE(
```

Comment:
There probably are two issues on this:
1. if we need to perform the check, then we need to align the writer to have this aligned enforcement.
2. we probably don't need to make this check as you don't need to load the entire file in memory all at once. We can improve this part as a followup, using an input stream with a controlled buffer.